feat: parse supervisor RPC errors

07Vaishnavi-Singh opened the pull request:
Closes #15671

### Changes
- [x] Added InvalidInboxEntry enum with variants for specific validation failures:
  - [x] UnknownChain: When a message targets an unsupported chain ID
- [x] Added SafetyLevel enum to represent message safety categories:
- [x] Implemented from_json_rpc method to map supervisor error codes to specific variants:
   - [x] Code -320501 maps to UnknownChain
   - [x] Codes in range -320999 to -320900 with safety level references map to MinimumSafety

07Vaishnavi-Singh commented on crates/optimism/txpool/src/supervisor/errors.rs:
```diff
@@ -7,17 +9,137 @@ pub enum InteropTxValidatorError {
+        if let Some(colon_idx) = remainder.find(':') {
+            let num_part = &remainder[colon_idx + 1..].trim();
+            return num_part.parse::<u64>().ok();
+        }
```
length check added

emhane commented:
also review from @SozinM would be nice 🫡

emhane commented on crates/optimism/txpool/src/supervisor/errors.rs:
```diff
@@ -7,17 +9,148 @@ pub enum InteropTxValidatorError {
+                #[allow(clippy::manual_range_contains)]
+                let chain_id = extract_chain_id(message).unwrap_or(0);
+                return Self::InvalidEntry(InvalidInboxEntry::UnknownChain(chain_id));
+            } else if code >= -320999 &&
+                code <= -320900 &&
+                (message.contains("safety") || message.contains("level"))
+            {
```
what we want to do, is probably create a new enum `SupervisorErrorCode` which contains all the supervisor protocol errors as variants like
`enum SupervisorErrorCode { UnknownChain = -320501, .. }` and then derive the conversion from `i32` with https://docs.rs/derive_more/latest/derive_more/derive.TryFrom.html

07Vaishnavi-Singh replied:
yeah wrapping the error codes inside an enum is a better way of dealing with them

emhane commented on crates/optimism/txpool/src/supervisor/errors.rs:
```diff
@@ -7,17 +94,237 @@ pub enum InteropTxValidatorError {
+                #[allow(clippy::int_plus_one)]
+                if colon_idx + 1 <= remainder.len() {
+                    // Ensure there are characters after the colon
+                    let num_part = &remainder[colon_idx + 1..].trim();
+                    return num_part.parse::<u64>().ok();
+                }
+            }
+        }
+    }
+
+    // Fallback: look for any number in the message
+    #[allow(clippy::is_digit_ascii_radix)]
+    message
+        .split_whitespace()
+        .filter_map(|word| word.trim_matches(|c: char| !c.is_digit(10)).parse::<u64>().ok())
+        .next()
+}
+
+/// Extracts a safety level value from error message patterns
+fn extract_safety_level(message: &str, key_param: &str) -> Option<SafetyLevel> {
+    // Look for patterns like "got: 0" or "expected: 1" in the message
+    let patterns = [format!("{key_param}: "), format!("{key_param} ")];
+    for pattern in &patterns {
+        if let Some(pos) = message.find(pattern.as_str()) {
+            if pos + pattern.len() < message.len() {
+                // Ensure there are characters after the pattern
+                let value_part = &message[pos + pattern.len()..];
+                let value = value_part.chars().next().and_then(|c| c.to_digit(10));
+
+                return match value {
+                    Some(0) => Some(SafetyLevel::Unsafe),
+                    Some(1) => Some(SafetyLevel::Safe),
+                    Some(2) => Some(SafetyLevel::Finalized),
+                    _ => None,
+                };
+            }
+        }
+    }
+    None
+}
```
let's skip this, and not parse the rpc error message at all and just rely on the code. error code is less risk for bugs than string error messages which may be worded differently across implementations. differing wording in strings can also be a harder bug to spot than spotting if the wrong error code is used across implementations, the latter is fairly straight forward bug to spot.

emhane commented on crates/optimism/txpool/src/supervisor/errors.rs:
```diff
@@ -7,17 +94,237 @@ pub enum InteropTxValidatorError {
+/// Invalid inbox entry
+#[derive(thiserror::Error, Debug)]
+pub enum InvalidInboxEntry {
+    /// Message does not meet minimum safety level
+    #[error("message does not meet min safety level, got: {got}, expected: {expected}")]
+    MinimumSafety {
+        /// Actual level of the message
+        got: SafetyLevel,
+        /// Minimum acceptable level that was passed to supervisor
+        expected: SafetyLevel,
+    },
```
let's skip this variant, it belonged to the old interop spec which used method `check_ExecutingMessages` instead of the current `check_AccessList`

emhane commented on crates/optimism/txpool/src/supervisor/errors.rs:
```diff
@@ -7,17 +94,237 @@ pub enum InteropTxValidatorError {
+        expected: SafetyLevel,
+    },
+    /// Invalid chain
+    #[error("unsupported chain id: {0}")]
```
```suggestion
    #[error("unsupported chain id")]
```
we don't need to extract the chain id into this variant, i.e. we don't need to parse the error message

07Vaishnavi-Singh commented on crates/optimism/txpool/src/supervisor/errors.rs:
```diff
@@ -7,17 +94,237 @@ pub enum InteropTxValidatorError {
+    DataCorruption,
+}
+
+/// Safety level of an inbox entry.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum SafetyLevel {
+    /// Unsafe level, no safety guarantees
+    Unsafe = 0,
+    /// Safe level, provides safety guarantees
+    Safe = 1,
+    /// Finalized level, provides finality guarantees
+    Finalized = 2,
+}
+
+impl fmt::Display for SafetyLevel {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            #[allow(clippy::use_self)]
+            SafetyLevel::Unsafe => write!(f, "unsafe"),
+            #[allow(clippy::use_self)]
+            SafetyLevel::Safe => write!(f, "safe"),
+            #[allow(clippy::use_self)]
+            SafetyLevel::Finalized => write!(f, "finalized"),
+        }
+    }
+}
+
```
yep removed it. ty!

07Vaishnavi-Singh commented:
@emhane thanks for the suggestions, I have removed all the out-of-scope code and implemented other recommendations as well. lmk if it needs further refactoring :)

SozinM commented on crates/optimism/txpool/src/supervisor/errors.rs:
```diff
@@ -7,17 +89,157 @@ pub enum InteropTxValidatorError {
 
+/// Invalid inbox entry
+#[derive(thiserror::Error, Debug)]
+pub enum InvalidInboxEntry {
```
I think that a lot of errors doesn't correspond to the entries itself, and rather are supervisor errors
For example:
```
UninitializedChainDatabase
NoParentForFirstBlock
```
Also i don't see the error related to the https://specs.optimism.io/interop/supervisor.html#supervisor_checkaccesslist-contents in spec, do you why is that @emhane?

emhane replied:
nice thanks, will open spec issue!

yes, aware some of the errors will prove redundant latest when we've completed the rust impl of the supervisor, but already by viewing the metrics for returned error codes for a while we will likely observe which errors aren't used. that, or we get a confirmation from original authors of the go impl which errors were created only with intention to use internally, whichever happens first. it was correct to include the ones we suspected could be public protocol errors from the existing list of errors in the go implementation, in order to make incremental progress. will open a spec issue to review and remove redundant error codes so it's parked ready for later.

SozinM replied:
My point was that the name `InvalidInboxEntry` does not fit in here IMO (because most of the errors doesn't have anything to do with the entries) and maybe we need to rename the error

emhane replied:
let's keep the name and instead address this on specs end https://github.com/ethereum-optimism/specs/issues/686

SozinM commented on crates/optimism/txpool/src/supervisor/errors.rs:
```diff
@@ -1,4 +1,86 @@
+#[derive(Debug, Clone, Copy, PartialEq, Eq, derive_more::TryFrom)]
+#[repr(i64)]
+#[try_from(repr)]
+#[allow(unreachable_pub)]
```
```suggestion
#[expect(unreachable_pub)]
```

emhane replied:
we can also export it from crate root already now in my opinion

emhane commented on crates/optimism/txpool/src/supervisor/errors.rs:
```diff
@@ -7,17 +90,157 @@ pub enum InteropTxValidatorError {
+/// Invalid inbox entry
+#[derive(thiserror::Error, Debug)]
+pub enum InvalidInboxEntry {
+    /// Invalid chain
+    #[error("unsupported chain id")]
+    UnknownChain,
+    /// Data was skipped and is not available
+    #[error("data was skipped or pruned and is not available")]
+    SkippedData,
+    /// Database for the chain is not initialized
+    #[error("chain database is not initialized")]
+    UninitializedChainDatabase,
+    /// Conflicting data exists
+    #[error("conflicting data exists in the database")]
+    ConflictingData,
+    /// Data is ineffective (already known)
+    #[error("data is already known and didn't change anything")]
+    IneffectiveData,
+    /// Data is out of order
+    #[error("data is out of order (too old or new)")]
+    OutOfOrder,
+    /// Waiting for replacement block
+    #[error("waiting for replacement block before progress can be made")]
+    AwaitingReplacement,
+    /// Data is outside allowed scope
+    #[error("data access not allowed due to limited scope")]
+    OutOfScope,
+    /// Cannot get parent of first block
+    #[error("cannot get parent of first block in database")]
+    NoParentForFirstBlock,
+    /// Data is from the future
+    #[error("data is not yet available (from the future)")]
+    FutureData,
+    /// Data was missed
+    #[error("data may exist but was not found (possibly different revision)")]
+    MissedData,
+    /// Data corruption
+    #[error("underlying database has I/O issues or is corrupted")]
+    DataCorruption,
+}
```
came to conclusion that we should collapse this enum with `SupervisorErrorCode`. then whenever we would want to display the error code (which there really isn't a need for in op-reth anyway, op-reth just parses the error code, never emits it) we just use the `as isize` or `as i64` syntax.

so let's remove the `SupervisorErrorCode` type and set the variants of this type equal to the corresponding `i64` and add the `TryFrom` derive macro. no need to change the variant names case, leave them camel case pls.

07Vaishnavi-Singh replied:
@emhane if I remove the ```SupervisorErrorCode``` then there is not need to expose ```SupervisorErrorCode``` in ```crates/optimism/src/txpool/src/supervisor/mod.rs``` as suggested in this [fix](https://github.com/paradigmxyz/reth/pull/15944/files/e7b77ac3aaa1d7b16dd81bc518be03292cbadcd7..2ce6e39f1034902c0ead2ece329d94723d48b864#diff-b6cbd92a476fc795a43001b585c1e58c92a854595683722d124bf203f4936978)

emhane reviewed (changes requested):
wrapping this up soon, great job so far

07Vaishnavi-Singh commented on crates/optimism/txpool/src/supervisor/errors.rs:
```diff
@@ -1,4 +1,85 @@
+    /// the latest data.
+    /// (by being too old or new).
+    #[error("data is out of order (too old or new)")]
+    OutOfOrder = -320900,
+
+    /// Happens when we know for sure that a replacement block is needed before progress
+    /// can be made.
+    #[error("waiting for replacement block before progress can be made")]
+    AwaitingReplacement = -320901,
+
+    // -3211XX OUT_OF_RANGE errors
+    /// Happens when data is accessed, but access is not allowed, because of a limited
+    /// scope.
+    /// E.g. when limiting scope to L2 blocks derived from a specific subset of the L1
+    /// chain.
+    #[error("data access not allowed due to limited scope")]
+    OutOfScope = -321100,
+
+    // -3212XX UNIMPLEMENTED errors
+    /// Happens when you try to get the previous block of the first block.
+    /// E.g. when trying to determine the previous source block for the first L1 block
+    /// in the database.
+    #[error("cannot get parent of first block in database")]
+    NoParentForFirstBlock = -321200,
+
+    // -3214XX UNAVAILABLE errors
+    /// Happens when data is just not yet available.
+    #[error("data is not yet available (from the future)")]
+    FutureData = -321401,
+
+    // -3215XX DATA_LOSS errors
+    /// Happens when we search the DB, know the data may be there, but is not (e.g.
+    /// different revision).
+    #[error("data may exist but was not found (possibly different revision)")]
+    MissedData = -321500,
+
+    /// Happens when the underlying DB has some I/O issue.
+    #[error("underlying database has I/O issues or is corrupted")]
+    DataCorruption = -321501,
+}
```
that is not required..... removing it from the PR description as well

emhane reviewed (approved):
great work! thanks a lot
